pkg/controller/seed-controller-manager/etcdbackup/etcd_backup_controller.go

```diff
@@ -0,0 +1,348 @@
+	if backup.DeletionTimestamp != nil {
+		log.Debug("Cleaning up all backups")
+		return nil, wrapErrorMessage("error cleaning up all backups: %v", r.deleteAllBackups(ctx, log, backup))
+	}
+
+	backupName, err := r.currentlyPendingBackupName(backup, cluster)
+	if err != nil {
+		return nil, fmt.Errorf("can't determine backup schedule: %v", err)
+	}
+
+	if backupName != "" {
+		err := r.createBackup(ctx, log, backupName, backup, cluster)
+		if err != nil {
+			return nil, fmt.Errorf("error creating backup: %v", err)
+		}
+	}
+
+	if err := r.deleteBackupsUpToRemaining(ctx, log, backup, backup.GetKeptBackupsCount()); err != nil {
+		return nil, fmt.Errorf("error expiring old backups: %v", err)
+	}
+
+	reconcile, err := r.computeReconcileAfter(backup)
+	if err != nil {
+		// should not happen at this point because the schedule was already parsed successfully above
+		return nil, fmt.Errorf("error computing reconcile interval: %v", err)
+	}
+
+	return reconcile, nil
+}
+
+// return name of backup to be done right now, or "" if no backup needs to be done right now
+func (r *Reconciler) currentlyPendingBackupName(backup *kubermaticv1.EtcdBackup, cluster *kubermaticv1.Cluster) (string, error) {
+	prefix := backupFileNamePrefix(backup.Name, cluster.Name)
+
+	if backup.Spec.Schedule == "" {
+		// no schedule set => we need exactly one backup (if none was created yet)
+		if backup.Status.LastBackupTime == nil {
+			return prefix, nil
+		}
+		return "", nil
+	}
+
+	schedule, err := parseCronSchedule(backup.Spec.Schedule)
+	if err != nil {
+		return "", err
+	}
+
+	lastBackupTime := backup.Status.LastBackupTime
+	if lastBackupTime == nil {
+		lastBackupTime = &metav1.Time{Time: backup.CreationTimestamp.Time}
+	}
+
+	if r.clock.Now().After(schedule.Next(lastBackupTime.Time)) {
+		return fmt.Sprintf("%s-%s", prefix, r.clock.Now().Format("2006-01-02T15:04:05")), nil
+	}
+
+	return "", nil
+}
+
+func (r *Reconciler) computeReconcileAfter(backup *kubermaticv1.EtcdBackup) (*reconcile.Result, error) {
+	if backup.Spec.Schedule == "" {
+		// no schedule set => only one immediate backup, which is already created at this point => all done, no need to reschedule
+		return nil, nil
+	}
+
+	schedule, err := parseCronSchedule(backup.Spec.Schedule)
+	if err != nil {
+		return nil, err
+	}
+
+	lastBackupTime := backup.Status.LastBackupTime
+	if lastBackupTime == nil {
+		lastBackupTime = &metav1.Time{Time: r.clock.Now()}
+	}
+
+	durationToNextBackup := schedule.Next(lastBackupTime.Time).Sub(r.clock.Now())
+	if durationToNextBackup < 0 {
+		durationToNextBackup = 0
+	}
+	return &reconcile.Result{Requeue: true, RequeueAfter: durationToNextBackup}, nil
+}
+
+func (r *Reconciler) createBackup(ctx context.Context, log *zap.SugaredLogger, fileName string, backup *kubermaticv1.EtcdBackup, cluster *kubermaticv1.Cluster) error {
```

Comment:
As we discussed offline, we shouldn't do this inside the seed controller, instead we will keep using the external container pattern.

Reply:
Done (see PR description).